core/domain/question_domain.py

```diff
@@ -0,0 +1,145 @@
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Domain objects relating to question."""
+
+from constants import constants
+from core.platform import models
+
+import feconf
+import utils
+(question_models,) = models.Registry.import_models([models.NAMES.question])
+
+class Question(object):
+    """Domain object for a question.
+
+    Attributes:
+        question_id: str. The unique id of the question.
+        title: str. The title of the question.
+        question_data: dict. The data for the question data.
+        data_schema_version: int. The schema version for the data.
+        collection_id: str. ID of the collection.
+        language_code: str. The ISO 639-1 code for the language this
+           question is written in.
+    """
+
+    def __init__(self, question_id, title, question_data, data_schema_version,
+                 collection_id, language_code):
+        """Constructs a Question domain object.
+
+        Args:
+            question_id: str. The unique id of the question.
+            title: str. The title of the question.
+            question_data: dict. The data for the question data.
+            data_schema_version: int. The schema version for the data.
+            collection_id: str. ID of the collection.
+            language_code: str. The ISO 639-1 code for the language this
+               question is written in.
+        """
+        self.question_id = question_id
+        self.title = title
+        self.question_data = question_data
+        self.data_schema_version = data_schema_version
+        self.collection_id = collection_id
+        self.language_code = language_code
+
+    def to_dict(self):
+        """Returns a dict representing this Question domain object.
+
+        Returns:
+            A dict, mapping all instance of fields of Question instance.
+        """
+        return {
+            'question_id': self.question_id,
+            'title': self.title,
+            'question_data': self.question_data,
+            'data_schema_version': self.data_schema_version,
+            'collection_id': self.collection_id,
+            'language_code': self.language_code
+        }
+
+    def validate(self):
+        """Validates the Question domain object before it is saved."""
+
+        if not isinstance(self.question_id, basestring):
+            raise utils.ValidationError(
+                'Expected id to be a string, received %s' % self.question_id)
+
+        if not isinstance(self.title, basestring):
+            raise utils.ValidationError(
+                'Expected title to be a string, received %s' % self.title)
+
+        if not isinstance(self.question_data, dict):
```

Comment:
What's the internal structure of the question data dict, and where is the validation for it?

Reply:
According to doc the dict should represent exp_domain.State domain object.
So, we can use the validation of the state domain object itself right?

Reply:
Yes. Note that this needs to be fully validated + specified in docstrings.

Reply:
Ping -- this comment needs to be addressed prior to submission, and tests added for it.